### PR TITLE
[Merged by Bors] - perf: mark generic `simp` lemmas with low priority

### PR DESCRIPTION
**Pull request by j-loreaux**
This takes some of the most common `simp` lemmas for hom classes (e.g., `map_one`, `map_zero`, `map_add`, `map_mul`, etc.) and marks them with priority `low`. The motivation is that these lemmas, by necessity, have very weak discrimination tree keys, so they apply very widely. However, when they match but fail to apply (as in the `TCSynth.lean` test file; last entry), they force type class synthesis to traverse almost the entirety of the hom class hierarchy looking for a match. This is a significant expenditure, so avoiding it when necessary is useful.

This seems to give reasonable performance gains, and the fallout from the priority seems to be quite minimal. In particular, one can restore the old behavior locally just by adding the relevant lemmas to the list passed to `simp`, (e.g., `simp [map_zero]`). This is needed only very rarely across Mathlib.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by j-loreaux** [comment]
!bench

File: Mathlib/Analysis/SpecialFunctions/Bernstein.lean [hunk 1]
```diff
@@ -43,6 +43,7 @@ This result proves Weierstrass' theorem that polynomials are dense in `C([0,1],
 although we defer an abstract statement of this until later.
 -/
 
+set_option says.verify true
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Bernstein.lean:46]:
```suggestion
```

**Comment by j-loreaux** [comment]
aesop -1.967%
simp -0.526%
TC -0.918%
lint instructions -2.046%
instructions -310B

seems like a pretty clear win, especially given that there were only ~20 places across Mathlib that needed changing, and in each case, all that is necessary is to increase the priority of the lemma by adding it to the list explicitly. Moreover, 3 of these places were actually improvements (e.g., we didn't need to call `abel` in two places).

Moreover, it reduces the number of instance heartbeats necessary in a test file. (from ~8400 down to 10)

**Comment by dupuisf** [comment]
Very nice!

bors d+

**Comment by j-loreaux** [comment]
bors merge

**Pull request closed without merging**